### PR TITLE
Add backend method `run_workflow`

### PR DESCRIPTION
**Pull request by Egor-S**
A part of #299

* Add `run_workflow` to backend interface
* Implement `get_workflows` for `RemoteRepo`
* Refactor `dstack run` to use new interface
* Refactor `dstack.core.repo` as a module

Further improvements:

* Remove `Backend` from `Provider`
* Allow overriding workflow data
* Allow running workflow without yaml

Example: run a workflow from remote repo via SDK

```python
from dstack.backend.local import LocalBackend
from dstack.core.repo import RemoteRepo

repo = RemoteRepo(repo_url="https://github.com/dstackai/dstack-playground.git")
repo.repo_data.repo_branch = "main"
backend = LocalBackend(repo=repo, auto_init=True)

run_name, jobs = backend.run_workflow("hello")
print(run_name)

job_heads = backend.list_job_heads(run_name)
for event in backend.poll_logs(job_heads, start_time=0, attached=False):
    print(event.log_message)
```

### CODE REVIEW

File: cli/dstack/backend/base/__init__.py [hunk 1]
```diff
@@ -240,6 +243,37 @@ def get_configurator(cls) -> Configurator:
     def delete_workflow_cache(self, workflow_name: str):
         pass
 
+    def run_workflow(
+        self,
+        workflow_name: str,
+        tag_name: Optional[str] = None,
+        ssh_pub_key: Optional[str] = None,
+        args: Optional[argparse.Namespace] = None,
+    ) -> (str, List[Job]):
```
Comment (r4victor) [inline cli/dstack/backend/base/__init__.py]: Tuple[str, List[Job]]

File: cli/dstack/backend/base/__init__.py [hunk 2]
```diff
@@ -240,6 +243,37 @@ def get_configurator(cls) -> Configurator:
...
+        workflow = self.repo.get_workflows(credentials=self.get_repo_credentials()).get(
+            workflow_name, {}
+        )
+        if workflow:
+            provider = providers.load_provider(workflow["provider"])
+        elif workflow_name in providers.get_provider_names():
+            provider = providers.load_provider(workflow_name)
+            workflow_name = None
+        else:
+            raise NameError(f"No workflow or provider '{workflow_name}' is found")
```
Comment (r4victor) [inline cli/dstack/backend/base/__init__.py]: Errors are part of the API, and we should define our own errors. `NameError` has specific different semantics.

File: cli/dstack/core/repo/remote.py [hunk 3]
```diff
@@ -90,31 +68,14 @@ def path(self, sep: str = ".") -> str:
...
-    @property
-    def repo_id(self) -> str:
-        return self.repo_ref.repo_id
-
-    @property
-    def repo_user_id(self) -> str:
-        return self.repo_ref.repo_user_id
-
-    @property
-    def repo_spec(self) -> "RepoSpec":
```
Comment (r4victor) [inline cli/dstack/core/repo/remote.py:108]: Why dropping `repo_spec` in favor of `RepoSpec.from_repo()`? Seems like pointless refactoring...

File: cli/dstack/backend/base/__init__.py [hunk 4]
```diff
@@ -240,6 +243,37 @@ def get_configurator(cls) -> Configurator:
...
+
+        run_name = self.create_run()
+        provider.load(self, args, workflow_name, workflow, run_name, ssh_pub_key)
```
Comment (r4victor) [inline cli/dstack/backend/base/__init__.py]: Seems like it's not that hard to move deps resolution logic to backend and avoid passing backend to provider.

File: cli/dstack/core/repo/remote.py (hunk 3 above)
Comment (Egor-S) [reply cli/dstack/core/repo/remote.py:108]: It resolves the circular dependency: `Repo.repo_spec` and `RepoSpec.repo`

**Pull request merged**